### DISCUSSION TITLE
RVCop64 - Corruption along C64 memory access

### DISCUSSION THREAD

**Discussion by pottendo** [zeldin/OrangeCart#2] [RVCop64]
Hi Marcus and fellow hackers,

Thanks, Marcus, again for your previous replies to me - helped alot to get the cart somehow doing something!

For the readers here:
I noticed the with RVCop64, that the 'mem_test' in C64 address space show some 'KO'.
All tests of the cartridge mem show OK.
As suggested my Marcus, I built the bitstream with 80Mhz (resulting in 2 warnings instead of 1 - see [1]), however the litex-Bios-bootscreen shows 80Mhz, and the performance seems subjectively better (i.e. lesser KO's), but still some 'KO's happen occasionally, when testing mem-ranges between 0x0000 and 0x10000 (c64 memory). The machine is stable, so I don't believe it's bad RAM - SuperREU works stable (Tested with Sonic and some self written program).

Following the example (oled) I now wrote my first program for the RISCV which just accessing screen-mem, see [2]. 
Once it's running at full speed (as fast as the CoProc runs), the C64 won't even boot. However, slowing it a bit by adding a delay loop, it works, but still crashes from time to time or prints random characters somewhere on the screen and/or corrupts the c64 memory.

So I suspect, the mem-timing from RVCop64 is somehow off - stressing access may result in wrong addressing, which leads to corruptions somewhere in c64 memory -> crash the c64.

Is my assumption wrong, that the RVCop can access c64 mem without any sort of 'sync' and still keep the c64 stably running?

Any clues, what could be done? 

bye,
martin

[1]:  Warning: Max frequency for clock '$glbnet$platform_ecp5pll0_clkout0': 65.61 MHz (FAIL at 80.00 MHz)
[2]: test program:
`
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <irq.h>
#include <libbase/uart.h>
#include <libbase/i2c.h>
#include <generated/csr.h>

volatile uint8_t *scr = 0x0420;
const int cnt = 300; // adjust to check: with 300 c64 crashes pretty soon, 3000 is already somwhow stable.

int main(void)
{
#ifdef CONFIG_CPU_HAS_INTERRUPT
	irq_setmask(0);
	irq_setie(1);
#endif
	uart_init();
	uint8_t c = 0;

	while (1)
	{
	    for (int i = 0; i < 256; i++)
	    {
		c = (c + 1) % 255;
		*(scr + i) = c;
		// without slowing the loop, it won't even boot the c64
		for (volatile int j = 0; j < cnt; j++)
		    ;
	    }
	    printf("hello world: %d.\n", c);
	}
	return 0;
}
`
// sorry, formatting the code may not have worked in this post

**Comment by zeldin** [comment]
Well, just from the first quick read-through of this, I can at least tell you that it's expected that the C64 will not boot if you hit the bus as fast as you can from the coprocessor.  Every bus cycle that you do through the expansion port is one that you steal from the 6510.  So if you do it all the time the poor CPU will not be able to fetch a single instruction, and thus be quite stuck until you let up.  😸 
Also, any write accesses you do during bootup could affect the KERNALs memory test, if you are unlucky.

**Reply by pottendo** [reply]
hi,
understood.
With a smaller delay loop (e.g. cnt set to 300 in my example), the system boots; at least in my few boot-cycles I haven't noticed a crash.
On my system, my program behaves comparable to the observations I've seen during 'mem-test <some C64 mem-range>'. Also there sometimes spurious characters appeared on the screen, then potentially showed 'KO's.
I don't think that these tests suffer from concurrent access to the same memory location in C64 mem. My case is just writing the screen-mem, the c64-cpu is executing the basic interpreter. Same for the mem-tests via RVTerm, I've tested locations like 0xc000 - 0xcfff or 0x2000-0x4000, which, I assume, are not used by anythin during RVTerm -> litex-bios cmds.
Of course the c64 executes its usual interrupts, so there is some activity asynchronosously happening, but not wrt. addressed mem-locations. 
Speculating here:
- Maybe the coproc-mem-access is confused by some interrupts on-going on the c64 and ends up in manipulating the wrong mem-location?
- Further cycle, or even sub-cycle alignement is needed to avoid addressing confusion between the CoProc and the host-memory. SuperREU doesn't seem to have a comparable issue, but is a different implementation, afaik.

In anyone could confirm that this is also happening on his system, I'd be happy to know. 
I'll continue to experiment and post some further observations here.

thanks for any feedback and/or ideas,
martin

**Reply by zeldin** [reply]
Yeah, for the areas 0xc000-0xcffff and 0x2000-0x4000 there should be no accesses (from interrupts or otherwise) by the 6510 once the initial memory test completes (i.e. after you see "30719 BASIC BYTES FREE" on screen).

If there is a sub-cycle timing issue it could indeed lead to contention even if the processors are not accessing the same area.  It is important that the two processors are not trying to drive bus signals at the same time or either could experience a read or write error.  Read/write accesses from the cartridge could also fail even if the 6510 has everything tri-stated, in case the timing requirements of the internal RAM are not met.

The implementation of the bus access logic is different between RVCop64 and SuperREU, but they are based on the same underlying algorithm, so they should behave the same.  A difference related to timing is that SuperREU runs with a system clock of 80MHz but RVCop64 normally uses 64MHz.  This is why I suggested to try RVCop64 at 80MHz.  When running both implementations at the same speed I think they should behave the same, so there is a perceived difference it could be related to how the interface is used.  When running SuperREU normally you are making monolithic transfers (unless you are running an application like the movie player which uses the SuperREU-specific pacing feature) where the 6510 is stopped once, all the data is transferred, and the the 6510 restarted.  With RVCop64 if you do accesses with a small delay, the 6510 gets restarted and stopped between each such access, meaning that if there is some stochastic failure linked to starting/stopping the 6510 it will be much more likely to happen.

If you have a good reproduction recipe it would be interesting to run it on some different systems like you suggest.  Personally I have not seen any failures when doing `mem_test` on the 0x2000-0x4000 region on my C64c.

**Reply by zeldin** [reply]
A slightly larger `mem_test` that should be safe to run is `mem_test 0x900 0x7700`.  This succeeds 20 times of 20 for me.

**Reply by pottendo** [reply]
hi,
I run RVCop64 with 80Mhz - as suggested - and my test-prog, with cnt set to 300 very quickly either shows some weird chars on the screen or even crashes the 6510.
You may go down with the value - even more stressing the byte-wise memory access.

I have a 'breadbin' C64, Assy No. 250425, date 1984, so more on the early side - unfortunately no other model around to double check.
I don't know if any major changes were made between these earlier models vs. the c64c redesign. I assume yes, but if this has any impact on this particular use-case, I can't tell.
So let me know in which direction I should experiment.

@all orangecart owners here: pls. share your experience! Thanx,

best regards, martin

**Comment by pottendo** [comment]
Dear Marcus, fellow hackers!

Got hold of another C64, Assy 250407, 1983 - so even older than my model.

Did some tests with Marcus' RVCop64 build (to rule out build-setup differences):
- the BIOS memtest:
```itex> mem_test 0x900 0x7700
Memtest at 0x900 (29.7KiB)...
  Write: 0x900-0x8000 29.7KiB
   Read: 0x900-0x8000 29.7KiB
Memtest OK
```
repeated several times, no NOKs!
- My mandelbrot test: https://photos.app.goo.gl/jvg25oZLDtcD4EfaA
- The testprogram from above, even without delay: https://photos.app.goo.gl/cdEEPxyhhKi4PpSP8

I conclude: my C64 seems to behave slightly different, incompatible with the OrangeCart. 
I'll let my C64 be checked by some friends, who have all the fancy test-harnesses... maybe the PLA or some RAM chip is a bit unsteady.
I may get a C64C as well; just to check another model (newer).

Let the community here know if you have experienced similar issues with your C64!

all the best, martin

**Comment by LeifBloomquist** [comment]
Hi, I'm getting similar errors here. Assy 250466.  Completely stock (NTSC) except for a Nano SwinSID, also using a modern PSU. OrangeCart is in a 3D printed case (from this wiki)

![image](https://user-images.githubusercontent.com/2716576/209743413-8b7ea41e-9479-4a5a-82f6-714f057ba167.png)

(The 64 actually locked after this command, though LiteX was still live after a reboot and re-entering RVTERM)

**Reply by pottendo** [reply]
hi,
once you've managed to get the Litex console with one of my bitstreams from [here](https://github.com/pottendo/RVCop64-pottendo), you may give it another try. These bitstreams use a slight differen `phase shift` parameter (15 instead of 10), which makes my breadbin C64 happy here.
Note that you have to respect the different memory mapping! (C64 mem is mapped to 0x0f000000):
```
litex> mem_test 0x0f000900 0x7700
Memtest at 0xf000900 (29.7KiB)...
  Write: 0xf000900-0xf008000 29.7KiB
   Read: 0xf000900-0xf008000 29.7KiB
Memtest OK
```
looking forward to get your feedback!
pottendo

**Comment by pottendo** [comment]
Hi,
here some update after a small test-series with two different C64s and various values for the `phase_shift` parameter set [here](https://github.com/zeldin/RVCop64/blob/e8f45ddb92637cdae2bf33fc8e2b4ef5e7cb05e1/hw/rtl/clockrecovery.py#L73).
The test was done by repeated around 20 times:
```
litex> mem_test 0x0f000900 0x7700
Memtest at 0xf000900 (29.7KiB)...
  Write: 0xf000900-0xf008000 29.7KiB
   Read: 0xf000900-0xf008000 29.7KiB
Memtest OK
```
(this is the `vexriscv_smp` bitstream, so C64 addresses are mapped to `0x0f000000`)
Note that this may vary when using the default clock-speed of 64Mhz - I always build for 80Mhz.

From this experiments I conclude that setting the parameter to `17` should be on the safe side to cover both models here with a 80Mhz RVCop64.
I may also get a 250469 version (which is @zeldin's) to reproduce figures again. 
Before doing the tests systematically I had to return the 250407 model, which I used the last weeks. This behaves like my 250425, though the exact boundaries may vary also here.

In case one needs some specific bitstream to try, let me know - I can build and send, if needed.

bye, pottendo

C64 Model | Softcore | Speed | Bitstream | Phaseshift 10 | Phaseshift 13 | Phaseshift 15 | Phaseshift 17 | Phaseshift 20 | Phaseshift 23
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
C64C, Assy 250466 | vexriscv_smp, FPU, 1 core | 80Mhz | Rvterm |   |   | ok |   |   |  
C64C, Assy 250466 | vexriscv_smp, FPU, 1 core | 80Mhz | uart=usb-acm |   |   | ok |   |   |  
C64C, Assy 250466 | vexriscv_smp, FPU, 1 core | 80Mhz | uart=usb-acm, uart2=serial |   |   |   | ok |   |  
C64C, Assy 250466 | vexriscv | 80Mhz | uart=usb-acm, uart2=serial | data errors: 8/7616 | ok | ok | ok | ok | ok
  |   |   |   |   |   |   |   |   |  
Assy 250425 | vexriscv | 80Mhz | uart=usb-acm, uart2=serial | data errors: 3365/7616 | ok | ok | ok | ok | ok
Assy 250425 | vexriscv_smp, FPU, 1 core | 80Mhz | uart=usb-acm, uart2=serial |   | ok |   | ok |   | data errors: 1/7616, not reproducible


PS: the reason why you see also some variations of the bitstream differing in the `uart...` configuration is that it looked to me that this had an influence on the mem-test, especially activating the second serial (uart2). But this didn't confirm in my tests later; as I had the results, I added it to the table...
